File: backend/schemas/order_line.py
"""
Pydantic schemas for OrderLine models - WideOrbit-compatible
"""

from pydantic import BaseModel, ConfigDict
from typing import Optional, Dict, Any
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID
from backend.models.order_line import RevenueType, SelloutClass


class OrderLineBase(BaseModel):
    """Base order line schema"""
    line_number: int
    start_date: date
    end_date: date
    station_id: UUID  # Required - each line targets one station
    station: Optional[str] = None  # Legacy field - kept for backward compatibility
    product: Optional[str] = None
    revenue_type: Optional[RevenueType] = None
    length: Optional[int] = None
    daypart: Optional[str] = None
    days_of_week: Optional[str] = None
    rate: Optional[Decimal] = None
    rate_type: Optional[str] = None
    sellout_class: Optional[SelloutClass] = None
    priority_code: Optional[str] = None
    spot_frequency: Optional[int] = None
    estimated_impressions: Optional[int] = None
    cpm: Optional[Decimal] = None
    cpp: Optional[Decimal] = None
    makegood_eligible: bool = True
    guaranteed_position: bool = False
    preemptible: bool = True
    inventory_class: Optional[str] = None
    product_code: Optional[str] = None
    deal_points: Optional[str] = None
    notes: Optional[str] = None


class OrderLineDigital(BaseModel):
    """Digital/streaming add-on fields"""
    platform: Optional[str] = None
    impressions_booked: Optional[int] = None
    delivery_window: Optional[str] = None
    targeting_parameters: Optional[Dict[str, Any]] = None
    companion_banners: Optional[Dict[str, Any]] = None


class OrderLineCreate(OrderLineBase, OrderLineDigital):
    """Schema for creating a new order line"""
    order_id: UUID


class OrderLineUpdate(BaseModel):
    """Schema for updating an order line"""
    line_number: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    station_id: Optional[UUID] = None
    station: Optional[str] = None  # Legacy field
    product: Optional[str] = None
    revenue_type: Optional[RevenueType] = None
    length: Optional[int] = None
    daypart: Optional[str] = None
    days_of_week: Optional[str] = None
    rate: Optional[Decimal] = None
    rate_type: Optional[str] = None
    sellout_class: Optional[SelloutClass] = None
    priority_code: Optional[str] = None
    spot_frequency: Optional[int] = None
    estimated_impressions: Optional[int] = None
    cpm: Optional[Decimal] = None
    cpp: Optional[Decimal] = None
    makegood_eligible: Optional[bool] = None
    guaranteed_position: Optional[bool] = None
    preemptible: Optional[bool] = None
    inventory_class: Optional[str] = None
    product_code: Optional[str] = None
    deal_points: Optional[str] = None
    notes: Optional[str] = None
    platform: Optional[str] = None
    impressions_booked: Optional[int] = None
    delivery_window: Optional[str] = None
    targeting_parameters: Optional[Dict[str, Any]] = None
    companion_banners: Optional[Dict[str, Any]] = None


class OrderLineResponse(OrderLineBase, OrderLineDigital):
    """Schema for order line response"""
    id: UUID
    order_id: UUID
    station_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: archive/librelog-web-api-original/src/main/java/com/onelpro/librelog/models/OrderLine.java
package com.onelpro.librelog.models;

import com.onelpro.librelog.enums.RevenueType;
import com.onelpro.librelog.enums.SelloutClass;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * Entity representing an order line in the system.
 */
@Entity
@Table(name = "order_lines", indexes = {
        @Index(name = "ix_order_lines_order_id", columnList = "order_id"),
        @Index(name = "ix_order_lines_start_date", columnList = "start_date"),
        @Index(name = "ix_order_lines_end_date", columnList = "end_date"),
        @Index(name = "ix_order_lines_station", columnList = "station"),
        @Index(name = "ix_order_lines_revenue_type", columnList = "revenue_type"),
        @Index(name = "ix_order_lines_daypart", columnList = "daypart"),
        @Index(name = "ix_order_lines_priority_code", columnList = "priority_code"),
        @Index(name = "ix_order_lines_makegood_eligible", columnList = "makegood_eligible"),
        @Index(name = "ix_order_lines_guaranteed_position", columnList = "guaranteed_position"),
        @Index(name = "ix_order_lines_preemptible", columnList = "preemptible"),
        @Index(name = "ix_order_lines_product_code", columnList = "product_code"),
        @Index(name = "ix_order_lines_station_id", columnList = "station_id")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class OrderLine {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id", nullable = false)
    private Order order;

    @Column(name = "line_number", nullable = false)
    private Integer lineNumber;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "end_date", nullable = false)
    private LocalDate endDate;

    @Column(name = "station", length = 100)
    private String station;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "station_id")
    private Station stationEntity;

    @Column(name = "product", length = 255)
    private String product;

    @Enumerated(EnumType.STRING)
    @Column(name = "revenue_type", columnDefinition = "revenuetype")
    private RevenueType revenueType;

    @Column(name = "length")
    private Integer length;

    @Column(name = "daypart", length = 50)
    private String daypart;

    @Column(name = "days_of_week", length = 7)
    private String daysOfWeek;

    @Column(name = "rate", precision = 12, scale = 2)
    private BigDecimal rate;

    @Column(name = "rate_type", length = 50)
    private String rateType;

    @Enumerated(EnumType.STRING)
    @Column(name = "sellout_class", columnDefinition = "selloutclass")
    private SelloutClass selloutClass;

    @Column(name = "priority_code", length = 50)
    private String priorityCode;

    @Column(name = "spot_frequency")
    private Integer spotFrequency;

    @Column(name = "estimated_impressions")
    private Integer estimatedImpressions;

    @Column(name = "cpm", precision = 10, scale = 2)
    private BigDecimal cpm;

    @Column(name = "cpp", precision = 10, scale = 2)
    private BigDecimal cpp;

    @Column(name = "makegood_eligible")
    @Builder.Default
    private Boolean makegoodEligible = true;

    @Column(name = "guaranteed_position")
    @Builder.Default
    private Boolean guaranteedPosition = false;

    @Column(name = "preemptible")
    @Builder.Default
    private Boolean preemptible = true;

    @Column(name = "inventory_class", length = 100)
    private String inventoryClass;

    @Column(name = "product_code", length = 100)
    private String productCode;

    @Column(name = "deal_points", columnDefinition = "TEXT")
    private String dealPoints;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "platform", length = 100)
    private String platform;

    @Column(name = "impressions_booked")
    private Integer impressionsBooked;

    @Column(name = "delivery_window", length = 100)
    private String deliveryWindow;

    @Column(name = "targeting_parameters", columnDefinition = "JSONB")
    private String targetingParameters;

    @Column(name = "companion_banners", columnDefinition = "JSONB")
    private String companionBanners;

    @Column(name = "created_at")
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;
}

